### DISCUSSION TITLE
Loop thru all msg segments and send a single csv file

### DISCUSSION THREAD

**Discussion by Stonypaul1** [nextgenhealthcare/connect#5207] [Q&A]
I am retrieving 13 rows of data from a SQL Database in a Database Reader and using a Channel Writer am sending it on to a subsequent channel with a Channel Reader source. In the Transformer of the Channel Reader I am using the following JavaScript to try and generate a single CSV file but it doesn't work - all I get is 13 separate csv files with just the 13 header titles in them but no data.
I had this working before when everything was self contained, so I got the 13 rows and added them to a data variable and used while(data.next()) { } successfully, but .next() won't work with msg nor will while(msg.children().next()) { }. 
I've tried for each(seg in msg.children()) and that gives me 13 csv files but this time with both the header and the data, but how do I create a single csv file with one set of headers and 13 rows of comma delimited data please? Thank you in advance.

```
const delimiter = ',';
const newLine = '\r\n';
var outputResultStr = new java.lang.StringBuilder();
var outputFileName = 'CSVData_' + DateUtil.getCurrentDate('yyyyMMDDHHMMSS');

function addRow(a,b,c,d,e,f,g,h,i,j, isHeader) {
	outputResultStr.append(a).append(delimiter)
				.append(b).append(delimiter)
				.append(c).append(delimiter)
				.append(d).append(delimiter)
				.append(e).append(delimiter)
				.append(f).append(delimiter)
				.append(g).append(delimiter)
				.append(h).append(delimiter)
				.append(i).append(delimiter)
				.append(j).append(isHeader ? newLine : '');					
}

addRow('Field01','Field02','Field03','Field04','Field05','Field06','Field07','Field08','Field09','Field10', true);

while(msg.next()) {
		addRow(msg['Field01'].toString(), msg['Field02'].toString(), msg['Field03'].toString(),msg['Field04'].toString(),
				msg['Field05'].toString(), msg['Field06'].toString(), msg['Field07'].toString(), msg['Field08'].toString(),
				msg['Field09'].toString(), msg['Field10'].toString(), false);
	}


msg = outputResultStr;
channelMap.put('final_output', outputFileName);
```

**Comment by joaryche** [comment]
perhaps using a globalChannelMap to keep concatenating to the final_output across messages until you are ready to write
or
combine the msg in the source channel using a batching structure and sending the whole data ("msg") as json or xml to the 2nd channel

**Reply by Stonypaul1** [reply]
> perhaps using a globalChannelMap to keep concatenating to the final_output across messages until you are ready to write or combine the msg in the source channel using a batching structure and sending the whole data ("msg") as json or xml to the 2nd channel

Thank you, both options sound doable, my Mirth knowledge is getting there but could you help me out with the syntax please?

**Reply by joaryche** [reply]
with the first ...
you would obtain the current value of the map
add the new row to the variable
then store the value in the map
    var value = globalChannelMap.get('final_output');
    var value += addrow(...);
    globalChannelMap.put('final_output', value);
the newer versions of Mirth®Connect have a link to the Users Guide in the bottom left task options.  Some good reading on the scope of each of the Map Collections.

the second way would need to have some separator in the data and then you could
turn batching on in the source pane of the source channel
go to the summary pane
and in the data types panel describe how the data needs to be batched

**Reply by Stonypaul1** [reply]
> with the first ... you would obtain the current value of the map add the new row to the variable then store the value in the map var value = globalChannelMap.get('final_output'); var value += addrow(...); globalChannelMap.put('final_output', value); the newer versions of Mirth®Connect have a link to the Users Guide in the bottom left task options. Some good reading on the scope of each of the Map Collections.
> 
> the second way would need to have some separator in the data and then you could turn batching on in the source pane of the source channel go to the summary pane and in the data types panel describe how the data needs to be batched

Thank you for taking the time to reply in detail and on a Saturday as well! 👍
The next time I'm at my laptop I'll give it a blast and let you know how I get on

**Comment by pacmano1** [comment]
What is the ultimate objective here?  Create a single CSV from a DB reader?

**Reply by pacmano1** [reply]
If yes, see https://forums.mirthproject.io/forum/mirth-connect/support/16869-?p=180027#post180027

**Reply by Stonypaul1** [reply]
> What is the ultimate objective here? Create a single CSV from a DB reader?

Yes pretty much, a single CSV file with multiple lines

**Reply by pacmano1** [reply]
The entire thing is explained there, even with a sample channel attached.

**Reply by Stonypaul1** [reply]
> The entire thing is explained there, even with a sample channel attached.

Ok thanks I'll check it out
